benchmarks/benchmark.py

```diff
@@ -0,0 +1,575 @@
+#!/usr/bin/python
+from __future__ import print_function
+from argparse import ArgumentParser, RawTextHelpFormatter
+import sys
+import subprocess
+import os
+
+sys.path.append("test")
+from sig_utils import *
+import itertools
+
+WORKING_FOLDER = "./benchmarks/"
+
+BENCHMARK_TEMPLATE = """
+static void {benchmark_name}(benchmark::State& state) {{
+{setup}
+  for (auto _ : state) {{
+{var_conversions}
+    auto start = std::chrono::high_resolution_clock::now();
+    
+{code}
+    auto end = std::chrono::high_resolution_clock::now();
+    auto elapsed_seconds =
+      std::chrono::duration_cast<std::chrono::duration<double>>(end - start);
+    state.SetIterationTime(elapsed_seconds.count());
+    stan::math::recover_memory();
+    benchmark::ClobberMemory();
+  }}
+}}
+BENCHMARK({benchmark_name})->RangeMultiplier({multi})->Range(1, {max_size})->UseManualTime();
+
+"""
+
+CUSTOM_MAIN = """
+int main(int argc, char** argv)
+{{
+  stan::math::ChainableStack::instance_->memalloc_.alloc({});
+  stan::math::recover_memory();
+  
+  ::benchmark::Initialize(&argc, argv);
+  ::benchmark::RunSpecifiedBenchmarks();
+}}
+"""
+
+overload_scalar = {
+    "Prim": "double",
+    "Rev": "stan::math::var",
+    "Fwd": "stan::math::fvar<double>",
+    "Mix": "stan::math::fvar<stan::math::var>",
+}
+
+
+def run_command(command):
+    """
+    Runs given command and waits until it finishes executing.
+    :param command: command to execute
+    """
+    print()
+    print(" ".join(command))
+    p1 = subprocess.run(command)
+    if p1.returncode != 0:
+        raise RuntimeError("command failed: " + command)
```

Comment:
at top

    from datetime import datetime

```suggestion
    current_time_utc = datetime.utcnow().strftime("%Y%m%d-%H%M%S.%f")
    stdout_path = os.path.join(HERE, "bench_stdout-{}.txt".format(current_time_utc))
    stderr_path = os.path.join(HERE, "bench_stderr-{}.txt".format(current_time_utc))
    with open(stdout, "wb") as f_out, open(stderr, "wb") as f_err:
        p1 = subprocess.run(command, stdout=f_out, stderr=f_err)
        if p1.returncode != 0:
            raise RuntimeError("command failed: " + command)
```

Reply:
I don't really want to redirect stdout and stderr. If user specifies appropriate arguments the results are saved anyway. At least I like to look at the output it produces to immediately see the results and to be able to guess how long it will take.

Reply:
Sound ok and there is always possibility to use `python ... > out.txt`